### PR TITLE
Make native libraries work in the pex package style.

### PR DESCRIPTION
**Pull request by mikekap**
This follows up (and fixes) using -rpath in python extensions to load the dependent libraries.

The one downside is that you can't simply load a library with `ctypes` by name - you have to
load it by full path (pkg_resources.resource_filename). I think this is acceptable since that's
"normal" in python (although definitely annoying).

### CODE REVIEW

**Comment by mikekap** [comment]
Seems the testrunner test is still angry: `//test/com/facebook/buck/testrunner:testrunner failed with exit code 137`

**Comment by sdwilsh** [comment]
@facebook-github-bot import

**Comment by mikekap** [comment]
small note: the hardcoded pkg_resources name in the test is required due to https://github.com/pantsbuild/pex/issues/210 .

**Comment by mikekap** [comment]
Ping

File: src/com/facebook/buck/python/CxxPythonExtensionDescription.java [hunk 1]
```diff
@@ -198,10 +200,16 @@ protected Path getExtensionPath(BuildTarget target, Flavor pythonPlatform, Flavo
                 args.platformLinkerFlags,
                 cxxPlatform)));
 
-    // Embed a origin-relative library path into the binary so it can find the shared libraries.
+    // Embed a extension-relative library path into the .so so it can find the shared libraries that
+    // it depends on.
+    String rootPath = MorePaths.relativize(
+        Paths.get(""), Paths.get(args.baseModule.or(""))).toString();
```
Comment (marcinkosiba) [inline src/com/facebook/buck/python/CxxPythonExtensionDescription.java]: Paths.get("") is a no-no, use the ProjectFileSystem (from params.getProjectFileSystem()) to resolve paths.

File: src/com/facebook/buck/cxx/CxxLinkableEnhancer.java [hunk 2]
```diff
@@ -163,6 +163,14 @@ public static CxxLink createCxxLinkableBuildRule(
     // the soname.
     if (linkType == Linker.LinkType.SHARED) {
       argsBuilder.add(new StringArg("-shared"));
+
```
Comment (marcinkosiba) [inline src/com/facebook/buck/cxx/CxxLinkableEnhancer.java]: won't this make all C++ libraries (even those built, say, for Android) reference that directory at runtime?

File: src/com/facebook/buck/python/CxxPythonExtensionDescription.java (hunk 1 above)
Comment (mikekap) [inline src/com/facebook/buck/python/CxxPythonExtensionDescription.java]: These aren't real paths - this attempts to get enough `../` so that a module placed in a directory specified in `args.baseModule` would find the "root".

File: src/com/facebook/buck/cxx/CxxLinkableEnhancer.java (hunk 2 above)
Comment (mikekap) [inline src/com/facebook/buck/cxx/CxxLinkableEnhancer.java]: Yes this makes all shared libraries relocatable without using LD_LIBRARY_PATH. Although I think android ignores rpaths anyway.

**Comment by mikekap** [comment]
Ping

File: src/com/facebook/buck/python/CxxPythonExtensionDescription.java [hunk 3]
```diff
@@ -201,12 +203,16 @@ protected Path getExtensionPath(BuildTarget target, Flavor pythonPlatform, Flavo
...
+    // it depends on.
+    Path baseModule = PythonUtil.getBasePath(params.getBuildTarget(), args.baseModule);
+    String rootPath = MorePaths.relativize(Paths.get(""), baseModule).toString();
     argsBuilder.addAll(
         StringArg.from(
             Linkers.iXlinker(
                 "-rpath",
-                String.format("%s/", cxxPlatform.getLd().resolve(ruleResolver).libOrigin()))));
+                String.format("%s/%s", cxxPlatform.getLd().resolve(ruleResolver).libOrigin(),
+                rootPath))));
```
Comment (Coneko) [inline src/com/facebook/buck/python/CxxPythonExtensionDescription.java]: The indentation makes it look like `rootPath` is the third argument of `Linkers.iXlinker` instead of `String.format`.

Move each argument to `String.format` on its own line.

File: test/com/facebook/buck/python/PythonBinaryIntegrationTest.java [hunk 4]
```diff
@@ -248,7 +243,9 @@ public void binaryIsCachedProperly() throws IOException {
 
   @Test
   public void externalPexToolAffectsRuleKey() throws IOException {
-    assumeThat(packageStyle, equalTo(PythonBuckConfig.PackageStyle.STANDALONE));
+    if (packageStyle != PythonBuckConfig.PackageStyle.STANDALONE) {
+      return;
+    }
```
Comment (Coneko) [inline test/com/facebook/buck/python/PythonBinaryIntegrationTest.java]: Why this change?

File: test/com/facebook/buck/python/PythonBinaryIntegrationTest.java [hunk 5]
```diff
@@ -275,7 +272,10 @@ public void externalPexToolAffectsRuleKey() throws IOException {
 
   @Test
   public void multiplePythonHomes() throws Exception {
-    assumeThat(Platform.detect(), not(Matchers.is(Platform.WINDOWS)));
+    if (Platform.detect() == Platform.WINDOWS) {
+      return;
+    }
+
```
Comment (Coneko) [inline test/com/facebook/buck/python/PythonBinaryIntegrationTest.java]: Why this change?

File: test/com/facebook/buck/python/PythonBinaryIntegrationTest.java (hunk 4 above)
Comment (mikekap) [inline test/com/facebook/buck/python/PythonBinaryIntegrationTest.java]: This (and below) are mostly cosmetic - it's to get rid of guaranteed assumption violations. In this case the pex tool isn't used for anything but the `STANDALONE` package style - on any platform/installation. It's not an assumption violation - it's just an artifact of parametrized tests that the `externalPexToolAffectsRuleKey[INPLACE,*]` test even exists. It boils down to giving a green build rather than a yellow one.

**Comment by facebook-github-bot-6** [comment]
@mikekap updated the pull request.

**Comment by mikekap** [comment]
ping

**Comment by facebook-github-bot-2** [comment]
@mikekap updated the pull request.

**Comment by Coneko** [comment]
@facebook-github-bot shipit

**Comment by mikekap** [comment]
ping

**Comment by Coneko** [comment]
This was blocked internally because it changes all libraries, not just the ones used in pexes.

cc @andrewjcg

**Comment by mikekap** [comment]
I could add a different `Linker.LinkType.RELOCATABLE_SHARED` instead. Would that be preferable?

**Comment by mikekap** [comment]
@Coneko @andrewjcg any word on what's the path forward here? Would you guys prefer to keep this out of the mainline tree?

**Comment by Coneko** [comment]
I'm not sure if @andrewjcg had any idea how to fix it, but the problem here is that the rpath is embedded in every DSO, regardless of whether we use it in a python binary with the pex package style or not.

**Comment by styurin** [comment]
@andrewjcg Do we want to merge this in?

**Comment by andrewjcg** [comment]
Yeah, I'm not completely comfortable about baking this in to *all* libraries.  Since the library deployment strategy is really binary/rule specific, ideally the top-level binary would set the RPATH, as is done with C/C++ binaries.  However, this is obviously tricky since the Python interpreter is prebuilt.  We had a similar issue with Lua binaries, so we ended up embedding the interpreter into a super simple C/C++ program and setting `-rpath` on that.  I wonder if we can do similar here (e.g. https://docs.python.org/3/extending/embedding.html#very-high-level-embedding).
